src/lib/keccak.ts

```diff
@@ -0,0 +1,217 @@
+  0x0000000080008009n,
+  0x000000008000000an,
+  0x000000008000808bn,
+  0x800000000000008bn,
+  0x8000000000008089n,
+  0x8000000000008003n,
+  0x8000000000008002n,
+  0x8000000000000080n,
+  0x000000000000800an,
+  0x800000008000000an,
+  0x8000000080008081n,
+  0x8000000000008080n,
+  0x0000000080000001n,
+  0x8000000080008008n,
+].map(Field.from);
+
+// Return a keccak state where all lanes are equal to 0
+const getKeccakStateZeros = (): Field[][] =>
+  Array.from(Array(KECCAK_DIM), (_) => Array(KECCAK_DIM).fill(Field.from(0)));
+
+// KECCAK HASH FUNCTION
+
+// ROUND TRANSFORMATION
+
+// First algrithm in the compression step of Keccak for 64-bit words.
+// C[x] = A[x,0] xor A[x,1] xor A[x,2] xor A[x,3] xor A[x,4]
+// D[x] = C[x-1] xor ROT(C[x+1], 1)
+// E[x,y] = A[x,y] xor D[x]
+// In the Keccak reference, it corresponds to the `theta` algorithm.
+// We use the first index of the state array as the x coordinate and the second index as the y coordinate.
+const theta = (state: Field[][]): Field[][] => {
+  const stateA = state;
+
+  // XOR the elements of each row together
+  // for all x in {0..4}: C[x] = A[x,0] xor A[x,1] xor A[x,2] xor A[x,3] xor A[x,4]
+  const stateC = stateA.map((row) =>
+    row.reduce((acc, next) => Gadgets.xor(acc, next, KECCAK_WORD))
+  );
+
+  // for all x in {0..4}: D[x] = C[x-1] xor ROT(C[x+1], 1)
+  const stateD = Array.from({ length: KECCAK_DIM }, (_, x) =>
+    Gadgets.xor(
+      stateC[(x + KECCAK_DIM - 1) % KECCAK_DIM],
+      // using (x + m mod m) to avoid negative values
+      Gadgets.rotate(stateC[(x + 1) % KECCAK_DIM], 1, 'left'),
+      KECCAK_WORD
+    )
+  );
+
+  // for all x in {0..4} and y in {0..4}: E[x,y] = A[x,y] xor D[x]
+  const stateE = stateA.map((row, index) =>
+    row.map((elem) => Gadgets.xor(elem, stateD[index], KECCAK_WORD))
+  );
+
+  return stateE;
+};
+
+// Second and third steps in the compression step of Keccak for 64-bit words.
+// B[y,2x+3y] = ROT(E[x,y], r[x,y])
+// which is equivalent to the `rho` algorithm followed by the `pi` algorithm in the Keccak reference as follows:
+// rho:
+// A[0,0] = a[0,0]
+// | x |  =  | 1 |
+// | y |  =  | 0 |
+// for t = 0 to 23 do
+//   A[x,y] = ROT(a[x,y], (t+1)(t+2)/2 mod 64)))
+//   | x |  =  | 0  1 |   | x |
+//   |   |  =  |      | * |   |
+//   | y |  =  | 2  3 |   | y |
+// end for
+// pi:
+// for x = 0 to 4 do
+//   for y = 0 to 4 do
+//     | X |  =  | 0  1 |   | x |
+//     |   |  =  |      | * |   |
+//     | Y |  =  | 2  3 |   | y |
+//     A[X,Y] = a[x,y]
+//   end for
+// end for
+// We use the first index of the state array as the x coordinate and the second index as the y coordinate.
+function piRho(state: Field[][]): Field[][] {
+  const stateE = state;
+  const stateB: Field[][] = getKeccakStateZeros();
+
+  // for all x in {0..4} and y in {0..4}: B[y,2x+3y] = ROT(E[x,y], r[x,y])
+  for (let x = 0; x < KECCAK_DIM; x++) {
+    for (let y = 0; y < KECCAK_DIM; y++) {
+      // No need to use mod since this is always positive
```

Comment:
This comments refers to an old function called mod which dealt with negative remainders, making them positive (mod) by adding the modulus. But this was removed, and the original comment referred to not needing to use mod and use just Ocaml's `%` as it was always positive. 

What I am saying is that this comment should be removed. It adds no meaning in this context.

Reply:
fun fact - we have a function called `mod` which handles negative inputs in our JS code base as well. I understood the comment that way :)